File: lib/python/tests/test_udfs.py
#  Copyright 2023 Commonwealth Scientific and Industrial Research
#  Organisation (CSIRO) ABN 41 687 119 230.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import os

from pyspark.sql import SparkSession, DataFrame
from pyspark.sql.functions import lit
from pyspark.sql.types import (
    StructType,
    StringType,
    StructField,
    BooleanType,
    Row,
    ArrayType,
    IntegerType,
)
from pytest import fixture

from pathling.coding import Coding
from pathling.functions import SNOMED_URI, to_snomed_coding
from pathling.udfs import (
    member_of,
    subsumes,
    subsumed_by,
    translate,
    display,
    PropertyType,
    Equivalence,
    property_of,
    designation,
)

PROJECT_DIR = os.path.abspath(
    os.path.join(os.path.dirname(__file__), os.pardir, os.pardir, os.pardir)
)


# noinspection PyProtectedMember
@fixture(scope="module")
def spark(pathling_ctx):
    return pathling_ctx.spark


CODING_TYPE = StructType(
    [
        StructField("id", StringType(), True),
        StructField("system", StringType(), True),
        StructField("version", StringType(), True),
        StructField("code", StringType(), True),
        StructField("display", StringType(), True),
        StructField("userSelected", BooleanType(), True),
        StructField("_fid", IntegerType(), True),
    ]
)

LOINC_URI = "http://loinc.org"

CodingRow = Row("id", "system", "version", "code", "display", "userSelected", "_fid")


def snomed_coding_row(code: str):
    return CodingRow(None, SNOMED_URI, None, code, None, None, None)


def loinc_coding_row(code: str):
    return CodingRow(None, LOINC_URI, None, code, None, None, None)


Result = Row("id", "result")


def test_member_of(spark: SparkSession):
    df = spark.createDataFrame(
        [
            ("code-1", snomed_coding_row("368529001")),
            ("code-2", loinc_coding_row("55915-3")),
            ("code-3", None),
        ],
        schema=StructType(
            [StructField("id", StringType()), StructField("code", CODING_TYPE)]
        ),
    )

    result_df_col = df.select(
        df["id"],
        member_of("code", "http://snomed.info/sct?fhir_vs=refset/723264001").alias(
            "is_member"
        ),
    )

    assert result_df_col.collect() == [
        Result("code-1", True),
        Result("code-2", False),
        Result("code-3", None),
    ]

    result_df_str = df.select(
        "id", member_of("code", "http://loinc.org/vs/LP14885-5").alias("is_member")
    )

    assert result_df_str.collect() == [
        Result("code-1", False),
        Result("code-2", True),
        Result("code-3", None),
    ]

    result_df_coding = df.limit(1).select(
        "id",
        member_of(
            Coding(SNOMED_URI, "368529001"),
            "http://snomed.info/sct?fhir_vs=refset/723264001",
        ).alias("result"),
    )

    assert result_df_coding.collect() == [Result("code-1", True)]


def test_member_of_array(spark: SparkSession):
    df = spark.createDataFrame(
        [
            (
                "code-1",
                [snomed_coding_row("368529001"), snomed_coding_row("368529002")],
            ),
            ("code-2", [loinc_coding_row("55915-3"), loinc_coding_row("55915-4")]),
            ("code-3", None),
        ],
        schema=StructType(
            [
                StructField("id", StringType()),
                StructField("code", ArrayType(CODING_TYPE)),
            ]
        ),
    )

    result_df_col = df.select(
        df["id"],
        member_of("code", "http://snomed.info/sct?fhir_vs=refset/723264001").alias(
            "is_member"
        ),
    )

    assert result_df_col.collect() == [
        Result("code-1", True),
        Result("code-2", False),
        Result("code-3", None),
    ]

    result_df_str = df.select(
        "id", member_of("code", "http://loinc.org/vs/LP14885-5").alias("is_member")
    )

    assert result_df_str.collect() == [
        Result("code-1", False),
        Result("code-2", True),
        Result("code-3", None),
    ]

    result_df_coding = df.limit(1).select(
        "id",
        member_of(
            Coding(SNOMED_URI, "368529001"),
            "http://snomed.info/sct?fhir_vs=refset/723264001",
        ).alias("result"),
    )

    assert result_df_coding.collect() == [Result("code-1", True)]


@fixture(scope="module")
def subsumption_df(spark: SparkSession):
    return spark.createDataFrame(
        [
            ("id-1", snomed_coding_row("107963000"), [snomed_coding_row("63816008")]),
            (
                "id-2",
                loinc_coding_row("55915-3"),
                [snomed_coding_row("63816008"), loinc_coding_row("55914-3")],
            ),
            ("id-3", None, [snomed_coding_row("107963000")]),
        ],
        schema=StructType(
            [
                StructField("id", StringType()),
                StructField("codeA", CODING_TYPE),
                StructField("codeB", ArrayType(CODING_TYPE)),
            ]
        ),
    )


def test_subsumes(subsumption_df):
    result_df = subsumption_df.select(
        "id", subsumes(subsumption_df["codeA"], "codeB").alias("result")
    )
    assert result_df.collect() == [
        Result("id-1", True),
        Result("id-2", False),
        Result("id-3", None),
    ]

    result_df = subsumption_df.select(
        "id", subsumes("codeA", Coding(SNOMED_URI, "63816008")).alias("result")
    )
    assert result_df.collect() == [
        Result("id-1", True),
        Result("id-2", False),
        Result("id-3", None),
    ]

    result_df = subsumption_df.select(
        "id",
        subsumes(Coding(LOINC_URI, "55914-3"), subsumption_df["codeB"]).alias("result"),
    )
    assert result_df.collect() == [
        Result("id-1", False),
        Result("id-2", True),
        Result("id-3", False),
    ]


def test_subsumed_by(subsumption_df):
    result_df = subsumption_df.select(
        "id", subsumed_by(subsumption_df["codeB"], "codeA").alias("result")
    )
    assert result_df.collect() == [
        Result("id-1", True),
        Result("id-2", False),
        Result("id-3", None),
    ]

    result_df = subsumption_df.select(
        "id", subsumed_by(Coding(SNOMED_URI, "63816008"), "codeA").alias("result")
    )
    assert result_df.collect() == [
        Result("id-1", True),
        Result("id-2", False),
        Result("id-3", None),
    ]

    result_df = subsumption_df.select(
        "id",
        subsumed_by(subsumption_df["codeB"], Coding(LOINC_URI, "55914-3")).alias(
            "result"
        ),
    )
    assert result_df.collect() == [
        Result("id-1", False),
        Result("id-2", True),
        Result("id-3", False),
    ]


def test_translate(spark: SparkSession):
    df = spark.createDataFrame(
        [
            ("id-1", snomed_coding_row("368529001")),
            ("id-2", loinc_coding_row("55915-3")),
            ("id-3", None),
        ],
        schema=StructType(
            [StructField("id", StringType()), StructField("code", CODING_TYPE)]
        ),
    )

    result_df = df.select(
        "id",
        translate(df["code"], "http://snomed.info/sct?fhir_cm=100").alias("result"),
    )

    assert result_df.collect() == [
        Result("id-1", [snomed_coding_row("368529002")]),
        Result("id-2", []),
        Result("id-3", None),
    ]

    result_df = df.select(
        "id",
        translate(
            "code",
            "http://snomed.info/sct?fhir_cm=100",
            equivalences={Equivalence.EQUIVALENT, Equivalence.RELATEDTO},
        ).alias("result"),
    )
    assert result_df.collect() == [
        Result("id-1", [snomed_coding_row("368529002"), loinc_coding_row("55916-3")]),
        Result("id-2", []),
        Result("id-3", None),
    ]

    result_df = df.select(
        "id",
        translate(
            "code",
            "http://snomed.info/sct?fhir_cm=100",
            equivalences={"equivalent", "relatedto"},
            target=LOINC_URI,
        ).alias("result"),
    )
    assert result_df.collect() == [
        Result("id-1", [loinc_coding_row("55916-3")]),
        Result("id-2", []),
        Result("id-3", None),
    ]

    result_df = df.select(
        "id",
        translate(
            "code",
            "http://snomed.info/sct?fhir_cm=200",
            equivalences=[Equivalence.EQUIVALENT, Equivalence.RELATEDTO],
        ).alias("result"),
    )
    assert result_df.collect() == [
        Result("id-1", []),
        Result("id-2", []),
        Result("id-3", None),
    ]

    result_df = df.limit(1).select(
        "id",
        translate(
            Coding(LOINC_URI, "55915-3"),
            "http://snomed.info/sct?fhir_cm=200",
            reverse=True,
            equivalences=Equivalence.RELATEDTO,
        ).alias("result"),
    )

    assert result_df.collect() == [
        Result("id-1", [snomed_coding_row("368529002")]),
    ]


@fixture(scope="module")
def property_df(spark: SparkSession):
    return spark.createDataFrame(
        [
            ("id-1", snomed_coding_row("439319006")),
            ("id-2", loinc_coding_row("55915-3")),
            ("id-3", None),
        ],
        schema=StructType(
            [StructField("id", StringType()), StructField("code", CODING_TYPE)]
        ),
    )


def test_display(property_df: DataFrame):
    expected_result = [
        Result("id-1", None),
        Result(
            "id-2",
            "Beta 2 globulin [Mass/volume] in Cerebral spinal fluid by Electrophoresis",
        ),
        Result("id-3", None),
    ]

    result_df = property_df.select("id", display("code").alias("result"))
    assert result_df.collect() == expected_result

    result_df = property_df.select("id", display(property_df["code"]).alias("result"))
    assert result_df.collect() == expected_result

    result_df = property_df.limit(1).select(
        "id", display(Coding(LOINC_URI, "55915-3")).alias("result")
    )
    assert result_df.collect() == [
        Result(
            "id-1",
            "Beta 2 globulin [Mass/volume] in Cerebral spinal fluid by Electrophoresis",
        )
    ]

    result_df = property_df.limit(1).select(
        "id", display(Coding(LOINC_URI, "55915-3"), "fr-FR").alias("result")
    )
    assert result_df.collect() == [
        Result("id-1", "Bêta-2 globulines [Masse/Volume] Liquide céphalorachidien")
    ]


def test_property_of(property_df: DataFrame):
    result_df = property_df.select(
        "id", property_of("code", "parent", PropertyType.CODE).alias("result")
    )
    assert result_df.collect() == [
        Result("id-1", ["785673007", "74754006"]),
        Result("id-2", []),
        Result("id-3", None),
    ]

    result_df = property_df.select(
        "id",
        property_of(property_df["code"], "inactive", PropertyType.BOOLEAN).alias(
            "result"
        ),
    )
    assert result_df.collect() == [
        Result("id-1", []),
        Result("id-2", [False]),
        Result("id-3", None),
    ]

    result_df = property_df.limit(1).select(
        "id", property_of(Coding(LOINC_URI, "55915-3"), "display").alias("result")
    )
    assert result_df.collect() == [
        Result(
            "id-1",
            [
                "Beta 2 globulin [Mass/volume] in Cerebral spinal fluid by Electrophoresis"
            ],
        )
    ]

    result_df = property_df.limit(1).select(
        "id",
        property_of(
            Coding(LOINC_URI, "55915-3"), "display", accept_language="de"
        ).alias("result"),
    )
    assert result_df.collect() == [
        Result(
            "id-1",
            [
                "Beta-2-Globulin [Masse/Volumen] in Zerebrospinalflüssigkeit mit Elektrophorese"
            ],
        )
    ]


USE_DISPLAY = Coding(
    "http://terminology.hl7.org/CodeSystem/designation-usage", "display"
)


def test_designation(property_df: DataFrame):
    result_df = property_df.select("id", designation("code").alias("result"))

    assert result_df.collect() == [
        Result(
            "id-1",
            [
                "Screening for phenothiazine in serum",
                "Screening for phenothiazine in serum (procedure)",
            ],
        ),
        Result(
            "id-2",
            [
                "Beta 2 globulin [Mass/volume] in Cerebral spinal fluid by Electrophoresis",
                "Bêta-2 globulines [Masse/Volume] Liquide céphalorachidien",
                "Beta 2 globulin:MCnc:Pt:CSF:Qn:Electrophoresis",
            ],
        ),
        Result("id-3", None),
    ]

    result_df = property_df.select(
        "id", designation("code", USE_DISPLAY).alias("result")
    )

    assert result_df.collect() == [
        Result("id-1", ["Screening for phenothiazine in serum"]),
        Result(
            "id-2",
            [
                "Beta 2 globulin [Mass/volume] in Cerebral spinal fluid by Electrophoresis",
                "Bêta-2 globulines [Masse/Volume] Liquide céphalorachidien",
            ],
        ),
        Result("id-3", None),
    ]

    result_df = property_df.select(
        "id", designation(property_df["code"], USE_DISPLAY, "fr-FR").alias("result")
    )
    assert result_df.collect() == [
        Result("id-1", []),
        Result("id-2", ["Bêta-2 globulines [Masse/Volume] Liquide céphalorachidien"]),
        Result("id-3", None),
    ]

    result_df = property_df.limit(1).select(
        "id",
        designation(
            Coding(SNOMED_URI, "439319006"),
            to_snomed_coding(lit("900000000000003001")),
            "en",
        ).alias("result"),
    )
    assert result_df.collect() == [
        Result("id-1", ["Screening for phenothiazine in serum (procedure)"])
    ]

File: lib/R/pathlyr/tests/testthat/test-udfs.R
def_spark <- function() {
  # Get the shaded JAR for testing purposes.
  spark <- sparklyr::spark_connect(master = "local[2]", config = list(
    #"spark.sql.warehouse.dir" = fs::dir_create_temp(),
    "spark.driver.memory" = "4g"
  ))

  #on.exit(sparklyr::spark_disconnect(spark), add = TRUE)
  spark
}


def_ptl_context <- function(spark) {

    encoders <- spark %>%
      invoke_static("au.csiro.pathling.encoders.FhirEncoders", "forR4") %>%
      invoke("withMaxNestingLevel", as.integer(0)) %>%
      invoke("withExtensionsEnabled", as.logical(FALSE)) %>%
      invoke("withOpenTypes", invoke_static(spark, "java.util.Collections", "emptySet")) %>%
      invoke("getOrCreate")

    terminology_service_factory <- spark %>%
       invoke_new("au.csiro.pathling.terminology.mock.MockTerminologyServiceFactory")

    spark %>%
      invoke_static("au.csiro.pathling.library.PathlingContext", "create",
                   spark_session(spark), encoders, terminology_service_factory)
}

LOINC_URI <- "http://loinc.org"


coding_row <- function(system, code) {
   coding <- list(
      x0_id = NA,
      x1_system = system ,
      x2_version = NA,
      x3_code = code,
      x4_display = NA,
      x5_userSelected = NA,
      x6__fid = NA
    )
  jsonlite::toJSON(
    as.list(coding),
    na = "null",
    auto_unbox = TRUE,
    digits = NA
  )
}

snomed_coding_row <- function(code) {
  coding_row(SNOMED_URI, code)
}

loinc_coding_row <- function(code) {
  coding_row(LOINC_URI, code)
}


snomed_coding_result <- function(code) {
  list(system = SNOMED_URI, code = code)
}


loinc_coding_result <- function(code) {
  list(system = LOINC_URI, code = code)
}


select_expr <-function(...) {
  mutate(..., .keep='none')
}

test_that("member_of", {
  spark <- def_spark()
  pc <- def_ptl_context(spark)

  df <- spark  %>% sdf_copy_to(
    tibble::tibble(
        id = c("code-1", "code-2", "code-3"),
        code = c(
          snomed_coding_row("368529001"),
          loinc_coding_row("55915-3"),
          NA)
      ), overwrite = TRUE, struct_columns = 'code'
    )

  result_df_col <- df %>%
    select_expr(
      id,
      is_member = !!trm_member_of(code, "http://snomed.info/sct?fhir_vs=refset/723264001"),
    )

  expect_equal(
    sdf_collect(result_df_col),
    tibble::tibble(
      id = c("code-1", "code-2", "code-3"),
      is_member = c(TRUE, FALSE, NA)
    )
  )

  result_df_str <- df %>%
    select_expr(
      id,
      is_member = !!trm_member_of(code, "http://loinc.org/vs/LP14885-5")
    )

  expect_equal(
    sdf_collect(result_df_str),
    tibble::tibble(
      id = c("code-1", "code-2", "code-3"),
      is_member = c(FALSE, TRUE, NA)
    )
  )

  result_df_coding <- df %>%
    head(1) %>%
    select_expr(
      id,
      result = !!trm_member_of(
        !!trm_to_snomed_coding("368529001"),
        "http://snomed.info/sct?fhir_vs=refset/723264001"
      )
    )

  expect_equal(
    sdf_collect(result_df_coding),
    tibble::tibble(
      id = "code-1",
      result = TRUE
    )
  )
})


test_that("translate", {
  spark <- def_spark()
  pc <- def_ptl_context(spark)

  df <- spark  %>% sdf_copy_to(
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      code = c(
        snomed_coding_row("368529001"),
        loinc_coding_row("55915-3"),
        NA)
    ), overwrite = TRUE, struct_columns = 'code'
  )

  result_df <- df %>%
    select_expr(
      id,
      result = !!trm_translate(code, "http://snomed.info/sct?fhir_cm=100")
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      result = list(
        list(snomed_coding_result("368529002")),
        list(),
        NA
      )
    )
  )

  result_df <- df %>%
    select_expr(
      id,
      result = !!trm_translate(
        code,
        "http://snomed.info/sct?fhir_cm=100",
        equivalences = c("equivalent", "relatedto")
      )
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      result = list(
        list(snomed_coding_result("368529002"), loinc_coding_result("55916-3")),
        list(),
        NA
      )
    )
  )

  result_df <- df %>%
    select_expr(
      id,
      result = !!trm_translate(
        code,
        "http://snomed.info/sct?fhir_cm=100",
        equivalences = c("equivalent", "relatedto"),
        target = LOINC_URI
      )
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      result = list(
        list(loinc_coding_result("55916-3")),
        list(),
        NA
      )
    )
  )

  result_df <- df %>%
    select_expr(
      id,
      result = !!trm_translate(
        code,
        "http://snomed.info/sct?fhir_cm=200",
        equivalences = c("equivalent", "relatedto")
      )
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      result = list(
        list(),
        list(),
        NA
      )
    )
  )

  result_df <- df %>%
    head(1) %>%
    select_expr(
      id,
      result = !!trm_translate(
        !!trm_to_coding("55915-3", LOINC_URI),
        "http://snomed.info/sct?fhir_cm=200",
        reverse = TRUE,
        equivalences = c("relatedto")
      )
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = "id-1",
      result = list(
        list(snomed_coding_result("368529002"))
      )
    )
  )
})



property_df <- function(spark) {
  df <- spark  %>% sdf_copy_to(
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      code = c(
        snomed_coding_row("439319006"),
        loinc_coding_row("55915-3"),
        NA
      )
    ), overwrite = TRUE, struct_columns = 'code'
  )

  return(df)
}

test_that("display", {
  spark <- def_spark()
  pc <- def_ptl_context(spark)

  property_df <- property_df(spark)

  expected_result <-
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      result = c(NA,"Beta 2 globulin [Mass/volume] in Cerebral spinal fluid by Electrophoresis", NA),
    )


  result_df <- property_df %>%
    select_expr(
      id,
      result = !!trm_display(code)
    )

  expect_equal(
    sdf_collect(result_df),
    expected_result
  )

  result_df <- property_df %>%
    select_expr(
      id,
      result = !!trm_display(code)
    )

  expect_equal(
    sdf_collect(result_df),
    expected_result
  )

  result_df <- property_df %>%
    head(1) %>%
    select_expr(
      id,
      result = !!trm_display(!!trm_to_coding("55915-3", LOINC_URI))
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = "id-1",
      result = "Beta 2 globulin [Mass/volume] in Cerebral spinal fluid by Electrophoresis"
    )
  )

  result_df <- property_df %>%
    head(1) %>%
    select_expr(
      id,
      result = !!trm_display(!!trm_to_coding("55915-3", LOINC_URI), "fr-FR")
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = "id-1",
      result = "Bêta-2 globulines [Masse/Volume] Liquide céphalorachidien"
    )
  )
})

test_that("property_of", {
  spark <- def_spark()
  pc <- def_ptl_context(spark)

  property_df <- property_df(spark)

  result_df <- property_df %>%
    select_expr(
      id,
      result = !!trm_property_of(code, "parent", PropertyType$CODE)
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      result = list(
        list("785673007", "74754006"),
        list(),
        NA
      )
    )
  )

  result_df <- property_df %>%
    select_expr(
      id,
      result = !!trm_property_of(code, "inactive", PropertyType$BOOLEAN)
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      result = list(
        logical(0),
        c(FALSE),
        NA
      )
    )
  )

  result_df <- property_df %>%
    head(1) %>%
    select_expr(
      id,
      result = !!trm_property_of(
        !!trm_to_coding("55915-3", LOINC_URI),
        "display"
      )
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = "id-1",
      result = list(
        list("Beta 2 globulin [Mass/volume] in Cerebral spinal fluid by Electrophoresis")
      )
    )
  )

  result_df <- property_df %>%
    head(1) %>%
    select_expr(
      id,
      result = !!trm_property_of(
        !!trm_to_coding("55915-3", LOINC_URI),
        "display",
        accept_language = "de"
      )
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = "id-1",
      result = list(
        list("Beta-2-Globulin [Masse/Volumen] in Zerebrospinalflüssigkeit mit Elektrophorese")
      )
    )
  )
})

USE_DISPLAY <- trm_to_coding("display",
  "http://terminology.hl7.org/CodeSystem/designation-usage"
)

test_that("designation", {
  spark <- def_spark()
  pc <- def_ptl_context(spark)

  property_df <- property_df(spark)

  result_df <- property_df %>%
    select_expr(
      id,
      result = !!trm_designation(code)
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      result = list(
        list(
          "Screening for phenothiazine in serum",
          "Screening for phenothiazine in serum (procedure)"
        ),
        list(
          "Beta 2 globulin [Mass/volume] in Cerebral spinal fluid by Electrophoresis",
          "Bêta-2 globulines [Masse/Volume] Liquide céphalorachidien",
          "Beta 2 globulin:MCnc:Pt:CSF:Qn:Electrophoresis"
        ),
        NA
      )
    )
  )

  result_df <- property_df %>%
    select_expr(
      id,
      result = !!trm_designation(code, !!USE_DISPLAY)
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      result = list(
        list("Screening for phenothiazine in serum"),
        list(
          "Beta 2 globulin [Mass/volume] in Cerebral spinal fluid by Electrophoresis",
          "Bêta-2 globulines [Masse/Volume] Liquide céphalorachidien"
        ),
        NA
      )
    )
  )

  result_df <- property_df %>%
    select_expr(
      id,
      result = !!trm_designation(code, !!USE_DISPLAY, "fr-FR")
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      result = list(list(), list("Bêta-2 globulines [Masse/Volume] Liquide céphalorachidien"), NA)
    )
  )

  result_df <- property_df %>%
    head(1) %>%
    select_expr(
      id,
      result = !!trm_designation(
        !!trm_to_coding("439319006", SNOMED_URI),
        !!trm_to_coding("900000000000003001", SNOMED_URI),
        "en"
      )
    )

  expect_equal(
    sdf_collect(result_df),
    tibble::tibble(
      id = "id-1",
      result = list(list("Screening for phenothiazine in serum (procedure)"))
    )
  )
})




subsumption_df <- function(spark) {
  df <- spark  %>% sdf_copy_to(
    tibble::tibble(
      id = c("id-1", "id-2", "id-3"),
      codeA = c(
        snomed_coding_row("107963000"),
        loinc_coding_row("55915-3"),
        NA
      ),
      codeB = list(
        list(snomed_coding_row("63816008")),
        list(snomed_coding_row("63816008"), loinc_coding_row("55914-3")),
        list(snomed_coding_row("107963000"))
      )
    ), overwrite = TRUE, struct_columns = c('codeA', 'codeB')
  )

  return(df)
}

#
# TODO: Fix and enable
#
# test_that("subsumes", {
#   spark <- def_spark()
#   pc <- def_ptl_context(spark)
#
#   subsumption_df <- subsumption_df(spark)
#
#   result_df <- subsumption_df %>%
#     select_expr(
#       id,
#       result = !!trm_subsumes(codeA, codeB)
#     )
#
#   expect_equal(
#     sdf_collect(result_df),
#     tibble::tibble(
#       id = c("id-1", "id-2", "id-3"),
#       result = list(TRUE, FALSE, NA)
#     )
#   )
#
#   result_df <- subsumption_df %>%
#     select_expr(
#       id,
#       result = !!trm_subsumes(codeA, !!trm_to_coding("63816008", SNOMED_URI))
#     )
#
#   expect_equal(
#     sdf_collect(result_df),
#     tibble::tibble(
#       id = c("id-1", "id-2", "id-3"),
#       result = list(TRUE, FALSE, NA)
#     )
#   )
#
#   result_df <- subsumption_df %>%
#     select_expr(
#       id,
#       result = !!trm_subsumes(
#         !!trm_to_coding("55914-3", LOINC_URI),
#         codeB
#       )
#     )
#
#   expect_equal(
#     sdf_collect(result_df),
#     tibble::tibble(
#       id = c("id-1", "id-2", "id-3"),
#       result = list(FALSE, TRUE, FALSE)
#     )
#   )
# })
#
# test_that("subsumed_by", {
#   spark <- def_spark()
#   pc <- def_ptl_context(spark)
#
#   subsumption_df <- subsumption_df(spark)
#
#   result_df <- subsumption_df %>%
#     select_expr(
#       id,
#       result = !!trm_subsumed_by(codeB, codeA)
#     )
#
#   expect_equal(
#     sdf_collect(result_df),
#     tibble::tibble(
#       id = c("id-1", "id-2", "id-3"),
#       result = list(TRUE, FALSE, NA)
#     )
#   )
#
#   result_df <- subsumption_df %>%
#     select_expr(
#       id,
#       result = !!trm_subsumed_by(
#         !!trm_to_coding("63816008", SNOMED_URI),
#         codeA
#       )
#     )
#
#   expect_equal(
#     sdf_collect(result_df),
#     tibble::tibble(
#       id = c("id-1", "id-2", "id-3"),
#       result = list(TRUE, FALSE, NA)
#     )
#   )
#
#   result_df <- subsumption_df %>%
#     select_expr(
#       id,
#       result = !!trm_subsumed_by(
#         codeB,
#         !!trm_to_coding("55914-3", LOINC_URI)
#       )
#     )
#
#   expect_equal(
#     sdf_collect(result_df),
#     tibble::tibble(
#       id = c("id-1", "id-2", "id-3"),
#       result = list(FALSE, TRUE, FALSE)
#     )
#   )
# })




